irohad/main/config/config.hpp

```diff
@@ -0,0 +1,191 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IROHA_CONFIG_HPP_
+#define IROHA_CONFIG_HPP_
+
+#include <sstream>
+#include <string>
+
+namespace iroha {
+  namespace config {
+
+    /**
+     * @class Config
+     * @brief Configuration loader, storage and validator.
+     */
+    class Config {
+     protected:
+      /**
+        * @struct Config::Service
+        * @brief Network service.
+        */
+      struct Service {
+        std::string host;
+        uint16_t port;
+      };
+
+      /**
+       * @struct Config::AuthService
+       * @brief Network service with authentication.
+       */
+      struct AuthService : public Config::Service {
+        std::string username;
+        std::string password;
+      };
+
+     public:
+      /**
+       * @struct Config::BlockStore
+       * @brief Config for block storage.
+       */
+      struct BlockStorage {
+        std::string path;  ///< path to the block storage
+      };
+
+      /**
+       * @struct Config::Cryptography
+       * @brief Everything that is required for cryptography is here.
+       */
+      struct Cryptography {
+        std::string certificate;  ///< path to the certificate
+        std::string key;          ///< path to the private key
+      };
+
+      /**
+       * Config for Redis.
+       */
+      using Redis = Service;
+
+      /**
+       * @struct Config::Torii
+       * @brief Config for torii gateway.
+       */
+      struct Torii : public Service {
+        /**
+         * Reurns preformatted "listen address" for torii:
+         * host:port
+         * @return
+         */
+        std::string listenAddress() const noexcept {
+          return this->host + ":" + std::to_string(this->port);
+        }
+      };
+
+      /**
+       * @struct Config::Postgres
+       * @brief Postgres config defined here.
+       */
+      struct Postgres : public AuthService {
+        /**
+         * Returns preformatted "options":
+         * host=localhost port=$port user=$user password=$pwd
+         * @return
+         */
+        std::string options() const noexcept {
+          std::stringstream ss;
+
+          // host=localhost port=$port user=$user password=$pwd
+          ss << "host=" << this->host << " port=" << this->port
+             << " user=" << this->username << " password=" << this->password;
+
+          return ss.str();
+        }
+      };
+
+      /**
+       * @struct Config::BlockchainOptions
+       * @brief Control options, such as creation of new ledger (given genesis
+       * block), or something else.
+       */
+      struct BlockchainOptions {
+        std::string genesis_block;  ///< path to the file with "genesis block"
+      };
+
+      /**
+       * Returns true, if config is loaded, false otherwise.
+       * @return
+       */
+      bool loaded() const noexcept;
```

Comment:
Use case?

Reply:
In current implementation `load()` is executed in constructor. It is very much possible that descendant of `Config` will not do this. Then, users of `Config` interface must check if config is loaded and react somehow if it is not.